IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,468 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <Foundation/Foundation.h>
+#import "MSIDAccountCacheItem.h"
+#import "MSIDAccountItemSerializer.h"
+#import "MSIDAccountType.h"
+#import "MSIDCredentialCacheItem.h"
+#import "MSIDCredentialItemSerializer.h"
+#import "MSIDCredentialType.h"
+#import "MSIDError.h"
+#import "MSIDJsonSerializer.h"
+#import "MSIDLogger+Internal.h"
+#import "MSIDLogger.h"
+#import "MSIDMacKeychainTokenCache.h"
+#import "MSIDUserInformation.h"
+#import "NSString+MSIDExtensions.h"
+
+/**
+This Mac cache stores serialized account and credential objects in the macOS "login" Keychain.
+There are three types of items stored:
+  1) Secret shareable artifacts (SSO credentials: Refresh tokens, other global credentials)
+  2) Non-secret shareable artifacts (account metadata)
+  3) Secret non-shareable artifacts (access tokens, ID tokens)
+
+In addition to the basic account & credential properties, the following definitions are used below:
+<account_id>    : “<home_account_id>-<environment>”
+<credential_id> : “<credential_type>-<client_id>-<realm>”
+<access_group>  : e.g. "com.microsoft.officecache"
+
+
+Type 1 (Secret shareable artifacts) Keychain Item Attributes
+============================================================
+ATTRIBUTE         VALUE
+~~~~~~~~~         ~~~~~~~~~~~~~~~~~~~~~~~~
+kSecClass         kSecClassGenericPassword
+kSecAttrAccount   <access_group>
+kSecAttrService   “Microsoft Credentials”
+kSecValueData     JSON data (UTF8 encoded) – shared credentials objects (multiple credentials saved in one keychain
+item)
+
+Type 2 (Non-secret shareable artifacts) Keychain Item Attributes
+================================================================
+ATTRIBUTE         VALUE
+~~~~~~~~~         ~~~~~~~~~~~~~~~~~~~~~~~~
+kSecClass         kSecClassGenericPassword
+kSecAttrAccount   <home_account_id>-<environment>-<realm>
+kSecValueData     JSON data (UTF8 encoded) – account object
+
+Type 3 (Secret non-shareable artifacts) Keychain Item Attributes
+===============================================================
+ATTRIBUTE         VALUE
+~~~~~~~~~         ~~~~~~~~~~~~~~~~~~~~~~~~
+kSecClass         kSecClassGenericPassword
+kSecAttrAccount   <access_group>-<app_bundle_id>-<home_account_id>-<environment>
+kSecAttrService   <credential_id>-<target>
+kSetAttrGeneric   <credential_id>
+kSecAttrType      Numeric Value: 2001=Access Token 2002=Refresh Token (Phase 1) 2003=IdToken
+kSecValueData     JSON data (UTF8 encoded) – credential object
+
+
+Type 1 JSON Data Example:
```

Comment:
Is this type-1 only? If so, provide examples of the others, and interleave.

Reply:
The example is for type-1 only. I'll add other examples.